// @decorators/express/lib/src/middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorMiddlewareHandler = exports.ERROR_MIDDLEWARE = exports.middlewareHandler = void 0;
const di_1 = require("@decorators/di");
/**
 * Create request middleware handler that uses class or function provided as middleware
 *
 * @param {Type} middleware
 *
 * @returns {RequestHandler}
 */
function middlewareHandler(middleware) {
    return function (req, res, next) {
        try {
            return getMiddleware(middleware, [req, res, next]);
        }
        catch (error) {
            next(error);
        }
    };
}
exports.middlewareHandler = middlewareHandler;
/**
 * Error Middleware class registration DI token
 */
exports.ERROR_MIDDLEWARE = new di_1.InjectionToken('ERROR_MIDDLEWARE');
/**
 * Add error middleware to the app
 *
 * @returns {ErrorRequestHandler}
 */
function errorMiddlewareHandler() {
    return function (error, req, res, next) {
        try {
            return getMiddleware(exports.ERROR_MIDDLEWARE, [error, req, res, next]);
        }
        catch (_a) {
            next(error);
        }
    };
}
exports.errorMiddlewareHandler = errorMiddlewareHandler;
/**
 * Instantiate middleware and invoke it with arguments
 *
 * @param {InjectionToken | Type} middleware
 * @param {any[]} args
 */
function getMiddleware(middleware, args) {
    const next = args[args.length - 1]; // last parameter is always the next function
    let instance;
    try {
        // first, trying to get instance from the container
        instance = di_1.Container.get(middleware);
    }
    catch (_a) {
        try {
            // if container fails, trying to instantiate it
            instance = new middleware();
        }
        catch (_b) {
            // if instantiation fails, try to use it as is
            instance = middleware;
        }
    }
    // first, assuming that middleware is a class, try to use it,
    // otherwise use it as a function
    const result = instance.use ?
        instance.use.apply(instance, args) :
        instance.apply(instance, args);
    // if result of execution is a promise, add additional listener to catch error
    if (result instanceof Promise) {
        result.catch(e => next(e));
    }
    return result;
}
