// @decorators/socket/lib/src/decorators/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const meta_1 = require("../meta");
/**
 * Listener decorator factory, creates listener decorator
 *
 * @param {EventType} type Listener type: io or socket
 * @param {string} event Event name
 * @param {Type[]} middleware Event middleware
 */
function makeDecorator(type, event, middleware) {
    return (target, methodName, descriptor) => {
        const meta = meta_1.getMeta(target);
        meta.listeners.push({ event, type, middleware, methodName });
        return descriptor;
    };
}
/**
 * Register **connection** event
 *
 * @see **io.on('connection', fn)**
 *
 * @param {Type[]} middleware Event middleware
 */
function Connection(middleware) {
    return makeDecorator(meta_1.EventType.IO, 'connection', middleware);
}
exports.Connection = Connection;
/**
 * @alias {Connection}
 */
exports.Connect = Connection;
/**
 * Registers global event
 *
 * @see **io.on**
 *
 * @param {string} event
 * @param {Type[]} middleware Event middleware
 */
function GlobalEvent(event, middleware) {
    return makeDecorator(meta_1.EventType.IO, event, middleware);
}
exports.GlobalEvent = GlobalEvent;
/**
 * Registers **disconnect** listener
 *
 * @see **socket.on('disconnect', fn)**
 *
 * @param {Type[]} middleware Event middleware
 */
function Disconnect(middleware) {
    return makeDecorator(meta_1.EventType.Socket, 'disconnect', middleware);
}
exports.Disconnect = Disconnect;
/**
 * Registers socket event
 *
 * @see **socket.on**
 *
 * @param {string} event
 * @param {Type[]} middleware Event middleware
 */
function Event(event, middleware) {
    return makeDecorator(meta_1.EventType.Socket, event, middleware);
}
exports.Event = Event;
;
