// @decorators/socket/lib/src/meta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Parameter types enum
 *
 * @export
 * @enum {number}
 */
var ParameterType;
(function (ParameterType) {
    ParameterType[ParameterType["IO"] = 0] = "IO";
    ParameterType[ParameterType["Socket"] = 1] = "Socket";
    ParameterType[ParameterType["Args"] = 2] = "Args";
    ParameterType[ParameterType["Ack"] = 3] = "Ack";
})(ParameterType = exports.ParameterType || (exports.ParameterType = {}));
/**
 * Event types
 *
 * @export
 * @enum {number}
 */
var EventType;
(function (EventType) {
    EventType[EventType["IO"] = 0] = "IO";
    EventType[EventType["Socket"] = 1] = "Socket";
})(EventType = exports.EventType || (exports.EventType = {}));
/**
 * Metadata class
 *
 * @export
 * @class Meta
 */
class SocketMeta {
    constructor() {
        /**
         * Middleware
         */
        this.middleware = [];
        /**
         * Event listeners
         */
        this.listeners = [];
        /**
         * Event listener params
         */
        this.params = {};
    }
}
exports.SocketMeta = SocketMeta;
/**
 * Get or initiate metadata on target
 *
 * @param target
 *
 * @returns {SocketIOMeta}
 */
function getMeta(target) {
    if (!target.__socket_meta__) {
        target.__socket_meta__ = new SocketMeta();
    }
    return target.__socket_meta__;
}
exports.getMeta = getMeta;
