// @decorators/socket/lib/src/middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const di_1 = require("@decorators/di");
/**
 * IO Middleware class registration DI token
 */
exports.IO_MIDDLEWARE = new di_1.InjectionToken('IO_MIDDLEWARE');
/**
 * Create request middleware handler that uses class or function provided as middleware
 *
 * @export
 * @param {Type | InjectionToken} middleware
 *
 * @returns {RequestHandler}
 */
function middlewareHandler(middleware) {
    return function (...args) {
        const next = args[args.length - 1];
        let instance;
        try {
            instance = di_1.Container.get(middleware);
        }
        catch (_a) {
            try {
                instance = new middleware();
            }
            catch (_b) {
                instance = middleware;
            }
        }
        // first, assuming that middleware is a class, try to use it,
        // otherwise use it as a function
        const use = instance.use ?
            instance.use : instance;
        try {
            const result = use.apply(instance, args);
            // if result of execution is a promise, add additional listener to catch error
            if (result instanceof Promise) {
                result.catch(next);
            }
            return result;
        }
        catch (e) {
            return next(e);
        }
    };
}
exports.middlewareHandler = middlewareHandler;
/**
 * Loops through all registered middlewares
 *
 * @param {Type[]} middleware Array of middleware
 * @param {any[]} [args = []] Arguments to pass in
 *
 * @returns {Promise<*>}
 */
function executeMiddleware(middleware, args = []) {
    function iteratee(done, i = 0) {
        try {
            middlewareHandler(middleware[i])(...args, (err) => {
                if (err) {
                    return done(err);
                }
                if (i === middleware.length - 1) {
                    return done(null);
                }
                iteratee(done, ++i);
            });
        }
        catch (e) {
            done(e);
        }
    }
    return new Promise((resolve, reject) => {
        if (middleware === undefined || middleware.length === 0) {
            return resolve();
        }
        iteratee((err) => err ? reject(err) : resolve());
    });
}
exports.executeMiddleware = executeMiddleware;
